File: app/src/main/java/arun/com/chromer/activities/history/HistoryFragment.java
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.activities.history;

import android.database.Cursor;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.CardView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SwitchCompat;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;

import javax.inject.Inject;

import arun.com.chromer.R;
import arun.com.chromer.activities.common.BaseFragment;
import arun.com.chromer.activities.common.Snackable;
import arun.com.chromer.activities.settings.Preferences;
import arun.com.chromer.di.components.FragmentComponent;
import arun.com.chromer.util.Utils;
import butterknife.BindView;
import butterknife.OnClick;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

/**
 * Created by arunk on 07-04-2017.
 */
public class HistoryFragment extends BaseFragment<History.View, History.Presenter> implements History.View {
    @BindView(R.id.history_list)
    RecyclerView historyList;
    @BindView(R.id.swipe_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.fab)
    FloatingActionButton fab;
    @BindView(R.id.error)
    TextView error;
    @BindView(R.id.enable_history_subtitle)
    TextView enableHistorySubtitle;
    @BindView(R.id.incognito_switch)
    SwitchCompat incognitoSwitch;
    @BindView(R.id.enable_history_card)
    CardView enableHistoryCard;


    private HistoryAdapter historyAdapter;

    @Inject
    History.Presenter presenter;

    @Override
    public void snack(@NonNull String message) {
        ((Snackable) getActivity()).snack(message);
    }

    @Override
    public void snackLong(@NonNull String message) {
        ((Snackable) getActivity()).snackLong(message);
    }


    @Override
    public void loading(boolean loading) {
        swipeRefreshLayout.setRefreshing(loading);
    }

    @Override
    public void setCursor(@Nullable Cursor cursor) {
        historyList.postDelayed(() -> {
            if (isAdded()) {
                historyAdapter.setCursor(cursor);
                error.setVisibility(cursor == null || cursor.isClosed() || cursor.getCount() == 0 ? VISIBLE : GONE);
            }
        }, 100);
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_history;
    }

    @NonNull
    @Override
    public History.Presenter createPresenter() {
        return presenter;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        historyAdapter.cleanUp();
    }


    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
        historyList.setLayoutManager(linearLayoutManager);
        historyAdapter = new HistoryAdapter(linearLayoutManager);
        historyList.setAdapter(historyAdapter);
        historyList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                historyAdapter.onRangeChanged();
            }
        });

        swipeRefreshLayout.setColorSchemeColors(
                ContextCompat.getColor(getContext(), R.color.colorPrimary),
                ContextCompat.getColor(getContext(), R.color.accent));
        swipeRefreshLayout.setOnRefreshListener(() -> presenter.loadHistory());

        enableHistoryCard.setOnClickListener(v -> incognitoSwitch.performClick());
        incognitoSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> Preferences.get(getActivity()).incognitoMode(!isChecked));
        enableHistorySubtitle.setText(getFormattedMessage());

        final ItemTouchHelper.SimpleCallback swipeTouch = new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                presenter.deleteHistory(historyAdapter.getWebsiteAt(viewHolder.getAdapterPosition()));
            }
        };
        final ItemTouchHelper itemTouchHelper = new ItemTouchHelper(swipeTouch);
        itemTouchHelper.attachToRecyclerView(historyList);
    }

    private CharSequence getFormattedMessage() {
        final String provider = Preferences.get(getContext()).customTabApp();
        if (provider == null) {
            return getString(R.string.enable_history_subtitle);
        } else {
            return Utils.html(getActivity(), String.format(getString(R.string.enable_history_subtitle_custom_tab), Utils.getAppNameWithPackage(getActivity(), provider)));
        }
    }

    @Override
    protected void inject(FragmentComponent fragmentComponent) {
        fragmentComponent.inject(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.loadHistory();
        incognitoSwitch.setChecked(!Preferences.get(getActivity()).incognitoMode());
        getActivity().setTitle(R.string.title_history);
    }

    @OnClick(R.id.fab)
    public void onClearAllFabClick() {
        if (historyAdapter.getItemCount() != 0) {
            new MaterialDialog.Builder(getActivity())
                    .title(R.string.are_you_sure)
                    .content(R.string.history_deletion_confirmation_content)
                    .positiveText(android.R.string.yes)
                    .negativeText(android.R.string.no)
                    .onPositive((dialog, which) -> presenter.deleteAll(getContext()))
                    .show();
        }
    }
}

File: app/src/main/java/arun/com/chromer/history/HistoryFragment.kt
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.history

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.database.Cursor
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import arun.com.chromer.R
import arun.com.chromer.di.fragment.FragmentComponent
import arun.com.chromer.settings.Preferences
import arun.com.chromer.shared.base.Snackable
import arun.com.chromer.shared.base.fragment.BaseFragment
import arun.com.chromer.util.Utils
import butterknife.OnClick
import com.afollestad.materialdialogs.MaterialDialog
import kotlinx.android.synthetic.main.fragment_history.*
import javax.inject.Inject

/**
 * Created by arunk on 07-04-2017.
 */
class HistoryFragment : BaseFragment(), Snackable {
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    @Inject
    lateinit var preferences: Preferences

    private lateinit var historyAdapter: HistoryAdapter

    private var viewModel: HistoryFragmentViewModel? = null

    private val formattedMessage: CharSequence
        get() {
            val provider = preferences.customTabPackage()
            return if (provider == null) {
                getString(R.string.enable_history_subtitle)
            } else {
                Utils.html(activity!!, String.format(getString(R.string.enable_history_subtitle_custom_tab), Utils.getAppNameWithPackage(activity!!, provider)))
            }
        }

    override fun getLayoutRes(): Int {
        return R.layout.fragment_history
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun snack(message: String) {
        (activity as Snackable).snack(message)
    }

    override fun snackLong(message: String) {
        (activity as Snackable).snackLong(message)
    }

    fun loading(loading: Boolean) {
        swipeRefreshLayout.isRefreshing = loading
    }

    private fun setCursor(cursor: Cursor?) {
        historyList.postDelayed({
            if (isAdded) {
                historyAdapter.setCursor(cursor)
                error.visibility = if (cursor == null || cursor.isClosed || cursor.count == 0) VISIBLE else GONE
            }
        }, 100)
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) {
            activity?.setTitle(R.string.title_history)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupHistoryList()
        setupIncognitoSwitch()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(HistoryFragmentViewModel::class.java)
        observeViewModel()
    }

    private fun observeViewModel() {
        viewModel?.apply {
            initHistoryLoader()
            loadingLiveData.observe(this@HistoryFragment, Observer { loading(it!!) })
            historyCursorLiveData.observe(this@HistoryFragment, Observer {
                setCursor(it!!)
            })
        }
        loadHistory()
    }

    private fun setupIncognitoSwitch() {
        enableHistoryCard.setOnClickListener { incognitoSwitch.performClick() }
        incognitoSwitch.setOnCheckedChangeListener { _, isChecked -> preferences.incognitoMode(!isChecked) }
        enableHistorySubtitle.text = formattedMessage
    }

    private fun setupHistoryList() {
        val linearLayoutManager = LinearLayoutManager(activity)
        historyList.apply {
            historyList.layoutManager = linearLayoutManager
            historyAdapter = HistoryAdapter(activity!!, linearLayoutManager)
            adapter = historyAdapter
            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                    historyAdapter.onRangeChanged()
                }
            })
        }

        swipeRefreshLayout.apply {
            setColorSchemeColors(ContextCompat.getColor(context!!, R.color.colorPrimary), ContextCompat.getColor(context!!, R.color.accent))
            setOnRefreshListener { loadHistory() }
        }

        val swipeTouch = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                viewModel?.deleteHistory(historyAdapter.getItemAt(viewHolder.adapterPosition))
            }
        }
        val itemTouchHelper = ItemTouchHelper(swipeTouch)
        itemTouchHelper.attachToRecyclerView(historyList)
    }

    private fun loadHistory() {
        viewModel?.loadHistory()
    }

    override fun inject(fragmentComponent: FragmentComponent) {
        fragmentComponent.inject(this)
    }

    override fun onResume() {
        super.onResume()
        loadHistory()
        incognitoSwitch.isChecked = !preferences.incognitoMode()
    }

    @OnClick(R.id.fab)
    fun onClearAllFabClick() {
        if (historyAdapter.itemCount != 0) {
            MaterialDialog.Builder(activity!!)
                    .title(R.string.are_you_sure)
                    .content(R.string.history_deletion_confirmation_content)
                    .positiveText(android.R.string.yes)
                    .negativeText(android.R.string.no)
                    .onPositive { _, _ ->
                        viewModel?.deleteAll({ rows ->
                            if (isAdded) {
                                snack(String.format(context!!.getString(R.string.deleted_items), rows))
                            }
                        })
                    }.show()
        }
    }
}
